storage.go

```diff
@@ -1027,3 +1028,235 @@ func storeSlab(storage SlabStorage, slab Slab) error {
 	}
 	return nil
 }
+
+// FixLoadedBrokenReferences traverses loaded slabs and fixes broken references in maps.
+// A broken reference is a StorageID referencing a non-existent slab.
+// To fix a map containing broken references, this function replaces broken map with
+// empty map having the same StorageID and also removes all slabs in the old map.
+// Limitations:
+// - only fix broken references in map
+// - only traverse loaded slabs in deltas and cache
+// NOTE: The intended use case is to enable migration programs in onflow/flow-go to
+// fix broken references.  As of April 2024, only 10 registers in testnet (not mainnet)
+// were found to have broken references and they seem to have resulted from a bug
+// that was fixed 2 years ago by https://github.com/onflow/cadence/pull/1565.
+func (s *PersistentSlabStorage) FixLoadedBrokenReferences() ([]StorageID, error) {
+
+	// parentOf is used to find root slab from non-root slab.
+	// Broken reference can be in non-root slab, and we need StorageID of root slab
+	// to replace broken map by creating an empty new map with same StorageID.
+	parentOf := make(map[StorageID]StorageID)
+
+	getRootSlabID := func(id StorageID) StorageID {
+		for {
+			parentID, ok := parentOf[id]
+			if ok {
+				id = parentID
+			} else {
+				return id
+			}
+		}
+	}
+
+	hasBrokenReferenceInSlab := func(id StorageID, slab Slab) bool {
+		if slab == nil {
+			return false
+		}
+
+		var isMetaDataSlab bool
+
+		switch slab.(type) {
+		case *ArrayMetaDataSlab, *MapMetaDataSlab:
+			isMetaDataSlab = true
+		}
+
+		var foundBrokenRef bool
+		for _, childStorable := range slab.ChildStorables() {
+
+			storageIDStorable, ok := childStorable.(StorageIDStorable)
+			if !ok {
+				continue
+			}
+
+			childID := StorageID(storageIDStorable)
+
+			// Track parent-child relationship of root slabs and non-root slabs.
+			if isMetaDataSlab {
+				parentOf[childID] = id
+			}
+
+			if s.existIfLoaded(childID) {
```

Comment:
Don't know if this makes sense: We could skip this check if its already found to be broken?
```suggestion
			if foundBrokenRef || s.existIfLoaded(childID) {
```

Reply:
ah I see, makes sense. Thanks for the clarification!